### PR TITLE
 [FEATURE] Add metadata to QGIS projects 

### PR DESCRIPTION
**Pull request by nyalldawson**
Implements the first part of https://github.com/qgis/QGIS-Enhancement-Proposals/issues/115

Allows for detailed metadata to be set in project properties, including abstracts, keywords, contact details, links, etc.

Developed for Arpa Piemonte (Dipartimento Tematico Geologia e Dissesto) within ERIKUS project

TODO: 
~~- expression context variables for project metadata items~~
~~- include metadata in layout exports~~

### CODE REVIEW

**Comment by nirvn** [comment]
@nyalldawson , I'm wondering whether it'd be better to have a function to access project (and eventually layer) metadata instead of exposing those through variables, e.g. metadata_project('author'). 

That would stand quite nicely alongside a possible metadata_layer('layer_name','metadata_name') function.

**Comment by nirvn** [comment]
Just to be clear, there's indeed nothing wrong with making project metadata available through variables, but that ultimately wouldn't work for layers, in the context of accessing such information in e.g. a expression-based layout label item.

**Comment by nyalldawson** [comment]
@nirvn 

>  I'm wondering whether it'd be better to have a function to access project (and eventually layer) metadata instead of exposing those through variables, e.g. metadata_project('author').

> That would stand quite nicely alongside a possible metadata_layer('layer_name','metadata_name') function.

Probably, yes. But I'd still opt for some explicit variables for simplicity also (as is done in this PR) 

It would be very nice to be able to do something like:

    array_to_string(apply(@project_layers,metadata_layer(@element,'attribution'),', ')

(where @project_layers = array of layer ids for layers in project and apply = function to apply a sub-expression to all elements in an array (where @element is the current array element)

**Comment by nirvn** [comment]
@nyalldawson , nice. While we're dreaming about what's possible, I'd add a is_layer_visible call there:

`array_to_string(apply(@project_layers,if(is_layer_visible(@element),metadata_layer(@element,'attribution'), null),',')`


Nice :)

**Comment by nyalldawson** [comment]
@nirvn

Actually we'd probably want to get the list of layers from a layout map item directly...

**Comment by nirvn** [comment]
@nyalldawson , right. I was thinking it'd be nice for the text decoration to support insertion of expressions (the same way layout label items do ATM). We could do all sorts of cool things there, inc. having attribution text based on visibility of layer(s) within canvas extent, etc.

File: src/core/qgsproject.cpp [hunk 1]
```diff
@@ -511,6 +511,7 @@ void QgsProject::clear()
   mCustomVariables.clear();
   mMetadata = QgsProjectMetadata();
   mMetadata.setCreationDateTime( QDateTime::currentDateTime() );
+  mMetadata.setAuthor( QgsApplication::userFullName() );
```
Comment (nirvn) [inline src/core/qgsproject.cpp]: I'm -0 on this -- working in a sensitive environment, and working with other people in the same environment, I cannot recall the number of times someone had his/her name (or a user name that's easily linked to the person) attached to a word document or PDF without his/her knowledge.

I recon this isn't necessarily the most common scenario, just serving food for thoughts :)

Comment (nyalldawson) [reply src/core/qgsproject.cpp]: How about a hidden settings key which avoids this?

Comment (nirvn) [reply src/core/qgsproject.cpp]: @nyalldawson , sounds like a reasonable compromise.

**Comment by NathanW2** [comment]
Hidden option to turn on or off?

**Comment by nyalldawson** [comment]
> Hidden option to turn on or off?

Turn off I say

File: src/core/qgsproject.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgsproject.cpp]: Done

**Comment by nirvn** [comment]
@nyalldawson , thanks for incorporating this anonymity option.

**Comment by nirvn** [comment]
@nyalldawson , just tested the PR, nice work.

Few thoughts:
- The metadata tab in the project properties window could be better placed lower, to adopt the location found in the vector and raster properties windows
- I noticed that the current user name is added to author metadata when opening a pre-existing project file, I'm wondering whether the author file should be left untouched here
- Do we need a "type" field for project? If so, the default values listed in the drop down combo box are a bit out of context

Also, and that's not specific to this PR at all (and should not block this), it's a bit weird to have the contacts panel and links panel offering a totally different UX to enter items. I'd favor the contacts panel method (i.e. showing line entries for each parameter of a contact entry) is we decide to harmonize that.

**Comment by nyalldawson** [comment]
> The metadata tab in the project properties window could be better placed lower, to adopt the location found in the vector and raster properties windows

I actually disagree here - I'd rather see the layer one moved higher. In my opinion metadata is one of the "fundamental" aspects of a project/layer, so accordingly should be higher placed.

**Comment by nyalldawson** [comment]
> I noticed that the current user name is added to author metadata when opening a pre-existing project file, I'm wondering whether the author file should be left untouched here

**Comment by nyalldawson** [comment]
> Do we need a "type" field for project? If so, the default values listed in the drop down combo box are a bit out of context

Agreed - it's freeform, so I'll leave it -- but none of the existing ISO standard options really fit. "collectionSession" or "series" is closest... but I think we just leave it empty by default until some more fitting codes are added to the standard. https://geo-ide.noaa.gov/wiki/index.php?title=ISO_19115_and_19115-2_CodeList_Dictionaries#MD_ScopeCode describes some extra ones, where "product" or "document" could be good candidates, but we better wait till they are added to the standard first.

**Comment by nyalldawson** [comment]
> it's a bit weird to have the contacts panel and links panel offering a totally different UX to enter items. I'd favor the contacts panel method (i.e. showing line entries for each parameter of a contact entry) is we decide to harmonize that.

I don't really understand this comment - they do both use the list-based approach. It's just that the contact tab has the extra (singular) fields shown at the top

**Comment by nirvn** [comment]
> I actually disagree here - I'd rather see the layer one moved higher. In my opinion metadata is one of the "fundamental" aspects of a project/layer, so accordingly should be higher placed.

I'd be fine with that, makes sense.

> I don't really understand this comment - they do both use the list-based approach. It's just that the contact tab has the extra (singular) fields shown at the top

God, forget this, I was confused. All good.

**Comment by nirvn** [comment]
@nyalldawson , just tested the updated PR, IMHO ready to merge :tada:

**Comment by nyalldawson** [comment]
Thanks @nirvn!

**Pull request merged**